app/src/main/scala/org/alephium/explorer/service/FinalizerService.scala

```diff
@@ -0,0 +1,152 @@
+import slick.jdbc.PostgresProfile.api._
+
+import org.alephium.explorer.foldFutures
+import org.alephium.explorer.persistence._
+import org.alephium.explorer.persistence.model.AppState
+import org.alephium.explorer.persistence.schema.AppStateSchema
+import org.alephium.explorer.persistence.schema.CustomGetResult._
+import org.alephium.explorer.persistence.schema.CustomSetParameter._
+import org.alephium.util.{Duration, TimeStamp}
+
+/*
+ * Syncing mempool
+ */
+
+trait FinalizerService extends SyncService.BlockFlow
+
+object FinalizerService extends StrictLogging {
+
+  // scalastyle:off magic.number
+  val finalizationDuration: Duration = Duration.ofSecondsUnsafe(6500)
+  def finalizationTime: TimeStamp    = TimeStamp.now().minusUnsafe(finalizationDuration)
+  def rangeStep: Duration            = Duration.ofHoursUnsafe(24)
+  // scalastyle:on magic.number
+
+  def apply(syncPeriod: Duration, databaseConfig: DatabaseConfig[PostgresProfile])(
+      implicit executionContext: ExecutionContext): FinalizerService =
+    new Impl(syncPeriod, databaseConfig)
+
+  private class Impl(val syncPeriod: Duration, val databaseConfig: DatabaseConfig[PostgresProfile])(
+      implicit val executionContext: ExecutionContext)
+      extends FinalizerService
+      with DBRunner {
+
+    override def syncOnce(): Future[Unit] = {
+      logger.debug("Finalizing")
+      finalizeOutputs(databaseConfig)
+    }
+  }
+
+  def finalizeOutputs(databaseConfig: DatabaseConfig[PostgresProfile])(
+      implicit executionContext: ExecutionContext): Future[Unit] = {
+    DBRunner.run(databaseConfig)(getStartEndTime()).flatMap {
+      case Some((start, end)) =>
+        finalizeOutputsWith(start, end, rangeStep, databaseConfig)
+      case None =>
+        Future.successful(())
+    }
+  }
+
+  def finalizeOutputsWith(start: TimeStamp,
+                          end: TimeStamp,
+                          step: Duration,
+                          databaseConfig: DatabaseConfig[PostgresProfile])(
+      implicit executionContext: ExecutionContext): Future[Unit] = {
+    var i = 0
+    logger.debug(s"Updating outputs")
+    val timeRanges =
+      BlockFlowSyncService.buildTimestampRange(start, end, step)
+    foldFutures(timeRanges) {
+      case (from, to) =>
+        DBRunner
+          .run(databaseConfig)(
+            for {
+              nb <- updateOutputs(from, to)
+              _  <- updateAppState(to)
+            } yield nb
+          )
+          .map { nb =>
+            i = i + nb
+            logger.debug(s"$i outputs updated")
+          }
+    }.map(_ => logger.debug(s"Outputs updated"))
+  }
+
+  private def updateOutputs(from: TimeStamp, to: TimeStamp): DBActionR[Int] = {
+    sqlu"""
+      UPDATE outputs o
+      SET spent_finalized = i.tx_hash
+      FROM inputs i
+      WHERE i.output_ref_key = o.key
+      AND o.main_chain=true
+      AND i.main_chain=true
+      AND i.block_timestamp >= $from
+      AND i.block_timestamp < $to;
```

Comment:
Just to confirm: is this using `PreparedStatement`?

Reply:
Yes this will use `PreparedStatement`. 

```sql
(statement_name: S_1,
prepare_time:    2022-04-21 19:55:04.896275+10,
parameter_types: {bigint,bigint},
statement:       UPDATE outputs o
      SET spent_finalized = i.tx_hash
      FROM inputs i
      WHERE i.output_ref_key = o.key
      AND o.main_chain=true
      AND i.main_chain=true
      AND i.block_timestamp >= $1
      AND i.block_timestamp < $2)
```